8주차 과제 - 레스토랑 앱 꾸미기

sejineeee opened the pull request:
## 내용
안녕하세요! 드디어 마지막 주네요😂
이번 주도 열심히 해보도록 하겠습니다!!!!
일단 slice 파일에 reducer에 작성했던 action들과 actions 파일에 있던 action creator를 추가해주었습니다.
추가적으로 작업한 이후, emotion을 이용하여, 앱을 꾸며보도록 하겠습니다!!
이번 한 주도 잘 부탁드립니다!!

sejineeee committed:
chore: emotion과 redux toolkit을 사용하기 위해서 설치

sejineeee committed:
createSlice를 사용하여 action creator와 reducer, action을 slice 파일에 분리

redux-toolkit 라이브러리에 createSlice를 사용
action creator와 reducer, actions를 slice 파일에 추가

sejineeee committed:
필요없는 파일 제거

sejineeee committed:
configureStore 사용, 그에 맞게 수정

sejineeee committed:
actions 파일 경로를 slice 파일 경로로 변경

sejineeee committed:
fix: reducer 객체로 전달

configureStore에 그냥 전달했던 reducer를 객체로 전달

daadaadaah reviewed (changes requested):
빠른 PR 너무 좋습니다! 마지막주~ 같이 힘내봐요! 👊
스타일링하시고 스샷도 함께 첨부 부탁드려요🙏

sejineeee committed:
refactor: redux toolkit 사용으로 인해서 reducer 수정

redux toolkit 사용으로 인해 reducer 수정
slice 폴더 변경으로 인한 경로 변경

sejineeee committed:
refactor: pages 폴더와 components 폴더로 각각의 관심사별로 분리

관심사별로 파일 분리
그에 따른 경로 변경

sejineeee committed:
refactor: 최적화를 위해서 React.memo 사용 및 사용을 위해 React import

sejineeee committed:
refactor: 최적화를 위해 handler 함수에 useCallback 사용

최적화를 위해 handler 함수에 useCallback 사용

sejineeee committed:
refactor: loginSlice로 loginPage 관련한 action, reducer 분리

loginSlice 파일에 loginPage 관련한 action, reducer 분리
requestLogin 비동기 함수 분리

sejineeee committed:
refactor: loginSlice 추가

loginSlice 추가

sejineeee commented:
## 내용
일단 pages 폴더와 components 폴더로 구분하였습니다. 이렇게 관심사별로 폴더 분리를 해보는 게 처음이라서 맞게 분리한 건지 잘 모르겠네요ㅎㅎㅎ
```plainText
📁src
 ㅏApp.jsx
 ㅏApp.test.jsx
 ㅏindex.js
 ㅏreducer.js
 ㅏreducer.test.js
 ㅏstore.js
 ㅏutils.js
 ㅏutils.test.js
 ㅏ📁components
    ㅏ📁Categories
        ㅏCategoriesContainer.jsx
        ㅏCategoriesContainer.test.jsx
    ㅏ📁MenuItems
        ㅏMenuItems.jsx 
        ㅏMenuItems.test.jsx
    ㅏ📁Regions
        ㅏRegionsContainer.jsx
        ㅏRegionsContainer.test.jsx
    ㅏ📁ReviewForm
        ㅏReviewForm.jsx
        ㅏReviewForm.test.jsx
    ㅏ📁Reviews
        ㅏReviews.jsx
        ㅏReviews.test.jsx
    ㅏ📁TextField
        ㅏTextField.jsx
        ㅏTextField.test.jsx
 ㅏ📁pages
    ㅏ📁AboutPage
        ㅏAboutPage.jsx
        ㅏAboutPage.test.jsx
    ㅏ📁HomePage
        ㅏHomePage.jsx
        ㅏHomePage.test.jsx
    ㅏ📁LoginPage
        ㅏLoginForm.jsx
        ㅏLoginForm.test.jsx
        ㅏLoginFormContainer.jsx
        ㅏLoginFormContainer.test.jsx
        ㅏLoginPage.jsx
        ㅏLoginPage.test.jsx
        ㅏLogoutForm.jsx
        ㅏLogoutForm.test.jsx
    ㅏ📁NotFoundPage
        ㅏNotFoundPage.jsx
        ㅏNotFoundPage.test.jsx
    ㅏ📁RestaurantPage
        ㅏRestaurantContainer.jsx
        ㅏRestaurantContainer.test.jsx
        ㅏRestaurantDetail.jsx
        ㅏRestaurantDetail.test.jsx
        ㅏRestaurantPage.jsx
        ㅏRestaurantPage.test.jsx
    ㅏ📁RestaurantsPage
        ㅏRestaurantsContainer.jsx
        ㅏRestaurantsContainer.test.jsx
        ㅏRestaurantsPage.jsx
        ㅏRestaurantsPage.test.jsx
 ㅏ📁services
 ㅏ📁slice
    ㅏloginSlice.js
    ㅏslice.js

```

아직 스타일 작업은 하지 않은 상태여서 스크린샷은 올리지 않았습니다!! 작업하는대로 스크린샷과 함께 올리도록 하겠습니다!!

daadaadaah reviewed (changes requested):
나눠보신 폴더 및 파일 구조 보고 궁금한 점이 있습니다!
어떤 Container 컴포넌트는 components 폴더에 있고, 어떤 Container 컴포넌트는 pages 폴더에 있는데, 어떤 기준으로 나누신건가요?ㅎㅎ

sejineeee commented:
> 나눠보신 폴더 및 파일 구조 보고 궁금한 점이 있습니다! 어떤 Container 컴포넌트는 components 폴더에 있고, 어떤 Container 컴포넌트는 pages 폴더에 있는데, 어떤 기준으로 나누신건가요?ㅎㅎ

categories, regions, menuItems, reviewForm, reviews, textField를 컴포넌트라고 생각을 하고, 나누다 보니 이런 파일 구조가 나왔습니다! 근데 제가 다시 보니... 관심사대로 제대로 분리가 되지 않은 것 같아서 다른 사람이 보면 혼동이 올 수 있을 거라고 생각이 드네요!
Page를 제외한 것들은 컴포넌트 폴더로 변경하도록 수정하겠습니다!

sejineeee committed:
fix: initialState 빈 객체에 초기값 추가

initialState에 빈 객체에 loginFields, accessToken 할당

sejineeee committed:
feat: slice 파일에 있던 review 관련 action을 분리

slice 파일에 있던 review 관련 action과 비동기 함수를 분리하여 추가

sejineeee committed:
feat: restaurant에 관한 action을 분리하여 추가

restaurant, restaurants에 관한 action을 해당 파일에 추가

sejineeee committed:
slice 파일 분리에 따라 reducer 객체에 추가

loginSlice, reviewSlice, restaurantSlice를 reducer에 추가

sejineeee committed:
refactor: pages 폴더 중 LoginPage 폴더에 있던 것을 components 폴더로 이동

sejineeee committed:
refactor: 파일 이동으로 경로 변경

sejineeee committed:
slice 파일 분리로 각 action, 비동기 함수에 대한 호출 경로 변경

sejineeee committed:
slice 파일 분리로dispatch type에 application을 loginSlice로 변경

sejineeee committed:
refactor: 폴더 관심사별로 재분류 및 그에 따른 경로 변경

pages 폴더에는 페이지 파일만 있도록 이동
components 폴더 하위에 Restaurant, RestaurantDetail, Restaurants 폴더 추가

sejineeee committed:
refactor: setAccessToken 호출 경로 변경 및 EatGo 텍스트 변경

setAccessToken 호출 경로 변경
레스토랑 웹페이지를 꾸미기 위해서 로고 텍스트 변경

sejineeee committed:
slice 분리로 인하여 필요 없어진 파일 삭제

sejineeee committed:
test: loginSlice 테스트 추가

기존 reducer test 파일에서 진행했던 login 관련 테스트를 loginSlice 테스트 파일에 추가

sejineeee committed:
test: restaurantSlice 테스트

기존 reducer 파일에서 진행되었던 restaurant 관련 테스트를
restaurantSlice 테스트에 추가

sejineeee committed:
test: reviewSlice 테스트

기존 reducer 테스트 파일에 있던 review 관련된 테스트를
reviewSlice 테스트 파일에 추가

sejineeee commented:
## 내용
남겨주신 코멘트를 보고, 다시 생각을 해보니, components와 page 관심사대로 분류되지 않았다고 생각이 들어서 수정하였습니다.
이렇게 폴더를 분류해본 것은 처음 하는 것이다 보니, 아직도 맞게 분류가 되었는지 확신이 잘 서지 않네요ㅎㅎㅎ
slice 폴더에는 loginSlice, restaurantSlice, reviewSlice 이렇게 로그인에는 로그인 관련된 것들, 레스토랑에는 레스토랑들에 관련된 것들, 리뷰에는 리뷰 관련된 reducer 들을 분리해보았습니다. 그에 맞게 테스트 파일들도 분리했습니다!

아직 스타일 작업은 들어가지 않은 상태여서 작업하는대로 바로바로 스크린샷과 함께 올리도록 하겠습니다!  🙌🏻

daadaadaah reviewed (changes requested):
이전에 나눠 보신 폴더구조보다 수정하신 폴더구조가 더 보기 편하네용!
다만, component 폴더 내에서 비즈니스로직을 담당하는 컴포넌트와 UI를 담당하는 컴포넌트가 섞여있는게 좀 아쉽긴해요! component 폴더 내에서 `containers` 랑  `presentational` 로 분리하는 것도 한가지 방법일 것 같아요!ㅎㅎ이제 스타일링 한번 도전해봅시당!👊

sejineeee committed:
fix: get 함수 매개변수 변경

state의 구조가 변경되었기 때문에 매개변수에 RORO 패턴 적용
sliceName 안에 있는 key를 가져올 수 있도록 수정

sejineeee committed:
test: get 함수에 대한 테스트 수정

get 함수 수정에 따른 테스트 수정

sejineeee committed:
test: useSelector selector 함수 반환값 변경

useSelector selector 됐을 때 state 반환 값을
각 해당하는 slice 객체가 감싸도록 변경

sejineeee committed:
fix: useSelector get 함수 인자 변경

utils get 함수의 구조 변경에 맞게 useSelector에서 호출한
get 함수에도 구조에 따라 인자 전달하도록 수정

sejineeee committed:
fix: 비동기 함수에 getState 호출한 곳에서도 구조 변경

sejineeee committed:
test: mockStore에 전달하는 인자 구조 변경

구조 변경으로 인해 mockStore에도 slice 객체가 한번 감싸도록 수정

sejineeee committed:
e2e test 통과를 위해 test 파일에 헤더를 'Eat Ate'로 변경

단순하게 헤더 보다는 다른 이름으로 하는 게 좋지 않을까 생각하여
Eat Ate로 바꿔보았음

sejineeee commented:
## 내용
<img width="780" alt="스크린샷 2022-12-02 오전 11 14 29" src="https://user-images.githubusercontent.com/86041335/205200396-30673f56-d145-48da-9eb4-dff56a2097d9.png">
slice 파일 분리 이후, 테스트 파일에서 에러가 나지 않길래... 처음에는 잘 동작하는 줄 알았다가, e2e test를 해보고 동작하지 않는다는 걸 알고, 테스트 파일도 수정을 같이 진행하였습니다!
근데 계속 이 에러가 나타나는데, RestaurantPage를 보면 accessToken을 사용한 부분이 없고, get 함수를 사용하지 않았는데, 어떻게 해결하면 좋을지 모르겠습니다.

daadaadaah commented on src/utils.test.js:
```diff
@@ -2,11 +2,13 @@ import { get, equal } from './utils';
 
-  const f = get('name');
-  const g = get('age');
+  const f = get({ sliceName: 'nameSlice', key: 'name' });
+  const g = get({ sliceName: 'nameSlice', key: 'age' });
```
get 유틸함수 이렇게 변경하신 이유가 궁금합니다!ㅎㅎ

sejineeee replied:
slice를 분리한 이후에 state 구조를 보니
```
loginSlice: { loginFields: { ... } ... },
restaurantSlice: { regions: { ... } ... }, 
reviewSlice: { reviewFields: { ... } ... }, 
```
이렇게 slice 안에 중첩된 구조여서 get 함수를 이렇게 바꾸게 되었습니다!

daadaadaah replied:
sliceName을 둔 이유가 궁금합니다!ㅎㅎ

sejineeee replied:
get을 하기 위해서 loginFields['accessToken'] 접근해야 돼서, sliceName을 get에 같이 전달하기 위해서 test에서 sliceName이라고 적은 것입니다! 혹시 두지 않고도 할 수 있는 방법이 있을까요...? 🥺

daadaadaah reviewed (changes requested):
> slice 파일 분리 이후, 테스트 파일에서 에러가 나지 않길래... 처음에는 잘 동작하는 줄 알았다가, e2e test를 해보고 동작하지 않는다는 걸 알고, 테스트 파일도 수정을 같이 진행하였습니다! 근데 계속 이 에러가 나타나는데, RestaurantPage를 보면 accessToken을 사용한 부분이 없고, get 함수를 사용하지 않았는데, 어떻게 해결하면 좋을지 모르겠습니다.

- RestaurantPage에서 무엇을 render를 하고 있으며, 그 컴포넌트에 어떤 상태들이 필요한지에 대해 살펴봅시다!

sejineeee committed:
chore: emotion-react 설치

Global Style 사용을 위해 emotion-react 설치

sejineeee committed:
head 태그에 구글 폰트 추가

sejineeee committed:
emotion Global style을 이용하여 reset style 설정

sejineeee committed:
header 스타일 적용

sejineeee committed:
HomePage에 nav 스타일 구현, SlideContainer 컴포넌트 호출

sejineeee committed:
test: 테스트에 필요한 상태 추가

restaurantPage에서 호출한 restaurantContainer에서 렌더링 되는 상태를
확인한 후에 필요한 상태를 테스트에 추가

sejineeee committed:
slide에 사용할 이미지 public폴더에 img 폴더를 만들어 추가

sejineeee committed:
feat: slideList 배열 생성

slide img의 src와 alt를 담은 slideList 배열 생성하여 export

sejineeee committed:
feat: SlideContainer 구현

slideList를 import 하여 map을 활용
Slide 컴포넌트를 호출하여 props에 src, alt 전달

sejineeee committed:
feat: Slide 컴포넌트 구현

src, alt를 props로 전달받아 사용
일단 img가 브라우저에 렌더링되는 것 확인

sejineeee commented:
## 내용
RestaurantPage에서 무엇이 렌더링되는지 확인한 후, 테스트 작업에 필요한 상태들을 추가해 수정하였습니다!
<img width="1468" alt="스크린샷 2022-12-03 오후 9 33 53" src="https://user-images.githubusercontent.com/86041335/205441291-d1724cf2-093d-4012-ac6f-b2d4054e757f.png">

일단 슬라이드 컴포넌트 만든 것까지는 올렸습니다!
추가적으로 현재 헤더 아래에 있는 사진들을 슬라이드 형태로 만들어서 올리도록 하겠습니다!

sejineeee committed:
LoginPage 스타일 적용

Login title 스타일, form과 title을 감싸고 있는 Container 스타일 구현

sejineeee committed:
loginForm 스타일 구현

label, input, button 스타일 적용

sejineeee committed:
logout button 스타일 구현

Logout button 스타일적용

daadaadaah commented on src/components/Slide/Slide.jsx:
```diff
@@ -0,0 +1,12 @@
+import styled from '@emotion/styled';
+
+const SlideItem = styled.img({
+  margin: '0 8px',
```
`px`, `rem`, `em` 단위에 대해 한번 공부해보세요!

daadaadaah reviewed (changes requested):
좋습니다! 남은 작업 마무리 해봅시당!ㅎㅎ스샷 감사합니당!! 스타일링 기대되네용😊

sejineeee commented:
## 내용
몸이 좋지 않아서 이후 작업이 진행되지 않았습니다ㅠㅠ
추후 조금 더 작업하여 올리고, 리뷰 반영하여 올리도록 하겠습니다!
이번 한 주도 감사합니다! 연말 잘 보내시길 바라겠습니다~

sejineeee committed:
chore: font-awesome 설치 및 모듈 설정

font-awesome 사용을 위해서 설치 및 webpack에 설정

sejineeee committed:
head 태그에 폰트 추가

sejineeee committed:
login page 스타일

sejineeee committed:
login form 스타일

sejineeee committed:
logout form 스타일

sejineeee committed:
categories, regions 스타일 구현

sejineeee committed:
homePage 스타일

sejineeee committed:
aboutPage 스타일

sejineeee committed:
restaurants 스타일 구현

sejineeee committed:
app 스타일

sejineeee committed:
font 추가

sejineeee committed:
singleQuote에서 dubbleQuote로 변경

sejineeee committed:
not found page 스타일

sejineeee committed:
homePage 스타일

sejineeee committed:
slide 구현

sejineeee committed:
slide 스타일

daadaadaah reviewed (changes requested):
세진님도 8주동안 수고 많으셨어요!👏👏👏👏
마지막 주에 컨디션이 안좋으신게 좀 아쉽긴하지만ㅠㅠ
코드숨에 배운 것들 습관화하셔서 좋은 개발자로 성장하시길 바랍니다~ 🚀
